### PR TITLE
[WIP] Add TSS (Thread Specific Storage) API in CPython 3.7+

### PR DESCRIPTION
**Pull request by sonots**
For https://github.com/cython/cython/issues/1932

### CODE REVIEW

**Comment by sonots** [comment]
I will try to create a compatibility layer separately.

**Comment by sonots** [comment]
I wrote compatibility layer codes as follows, but I am pretty not sure where I shout put this:


```c
#if PY_VERSION_HEX <= 0x030700A1
// TSS (Thread Specific Storage) API in CPython 3.7+

#define Py_tss_NEEDS_INIT 0
typedef int Py_tss_t;
int PyThread_tss_create(Py_tss_t *key) {
  // PyThread_create_key will report initial success,
  // but attempts to use the returned key are likely to fail.
  *key = PyThread_create_key();
  return 0;  // always success
}
Py_tss_t * PyThread_tss_alloc(void) {
  int *key = malloc(sizeof(Py_tss_t));
  *key = Py_tss_NEEDS_INIT;
  return key;
}
void PyThread_tss_free(Py_tss_t *key) {
  free(key);
}
int PyThread_tss_is_created(Py_tss_t *key) {
  return *key != Py_tss_NEEDS_INIT;
}
void PyThread_tss_delete(Py_tss_t *key) {
  PyThread_delete_key(key);
  *key = Py_tss_NEEDS_INIT;
}
int PyThread_tss_set(Py_tss_t *key, void *value) {
  return PyThread_set_key_value(key, value);
}
void * PyThread_tss_get(Py_tss_t *key) {
  return PyThread_get_key_value(key);
}
// PyThread_delete_key_value(key) is equalivalent to PyThread_set_key_value(key, NULL)
// PyThread_ReInitTLS() is a no-op

#endif
```

**Comment by scoder** [comment]
I would change the preprocessor guard to
`#if PY_VERSION_HEX <= 0x030700A1 && !defined(PyThread_tss_create)`
Eventually, PyPy might also support this, and they will usually have `#define`s for the C-API functions.

And then I think it should go right into the preamble section in `ModuleSetupCode.c`, maybe after the `PyThreadState` adaptation code near line 260. Note that all functions must be `static`.

Also look at the GIL handling code at the end of that file. It might need adaptation to use the new API if available.

**Comment by sonots** [comment]
I've added the compatible layer codes at the bottom of `ModuleSetupCode.c` because I wanted to use `CYTHON_INLINE` and it was defined at line 536.

I am still not sure how to use the codes. I wrote a `hello.pyx` as:

```
cdef Py_tss_t tss_key
if PyThread_tss_create(&tss_key):
    pass

cdef int the_value = 1;
if PyThread_tss_get(&tss_key) == NULL:
    PyThread_tss_set(&tss_key, <void *>&the_value)

PyThread_tss_delete(&tss_key);

cdef Py_tss_t *ptr_key = PyThread_tss_alloc()

PyThread_tss_free(ptr_key)
ptr_key = NULL;
```

Cythonizing succeeds, but compilation fails as:


```
hello.c:903:8: error: unknown type name ‘Py_tss_t’
 static Py_tss_t __pyx_v_5hello_tss_key;
        ^
hello.c:905:8: error: unknown type name ‘Py_tss_t’
 static Py_tss_t *__pyx_v_5hello_ptr_key;
        ^
hello.c: In function ‘PyInit_hello’:
hello.c:1131:16: warning: implicit declaration of function ‘PyThread_tss_create’ [-Wimplicit-function-declaration]
   __pyx_t_1 = (PyThread_tss_create((&__pyx_v_5hello_tss_key)) != 0);
                ^
hello.c:1151:17: warning: implicit declaration of function ‘PyThread_tss_get’ [-Wimplicit-function-declaration]
   __pyx_t_1 = ((PyThread_tss_get((&__pyx_v_5hello_tss_key)) == NULL) != 0);
                 ^
hello.c:1151:61: warning: comparison between pointer and integer
   __pyx_t_1 = ((PyThread_tss_get((&__pyx_v_5hello_tss_key)) == NULL) != 0);
                                                             ^
hello.c:1161:5: warning: implicit declaration of function ‘PyThread_tss_set’ [-Wimplicit-function-declaration]
     PyThread_tss_set((&__pyx_v_5hello_tss_key), ((void *)(&__pyx_v_5hello_the_value)));
     ^
hello.c:1179:3: warning: implicit declaration of function ‘PyThread_tss_delete’ [-Wimplicit-function-declaration]
   PyThread_tss_delete((&__pyx_v_5hello_tss_key));
   ^
hello.c:1188:28: warning: implicit declaration of function ‘PyThread_tss_alloc’ [-Wimplicit-function-declaration]
   __pyx_v_5hello_ptr_key = PyThread_tss_alloc();
                            ^
hello.c:1188:26: warning: assignment makes pointer from integer without a cast [-Wint-conversion]
   __pyx_v_5hello_ptr_key = PyThread_tss_alloc();
                          ^
hello.c:1196:3: warning: implicit declaration of function ‘PyThread_tss_free’ [-Wimplicit-function-declaration]
   PyThread_tss_free(__pyx_v_5hello_ptr_key);
```

**Comment by dalcinl** [comment]
@sonots Could you add `# cython: fast_gil=True` at the top of `hello.pyx` and try again?

**Comment by sonots** [comment]
@dalcinl I got same failure.

**Comment by scoder** [comment]
Ah, sorry. I didn't tell you that the utility code files are split into sections that are inserted explicitly by name.

I moved the Python compatibility code into its own section now, so that you have access to `CYTHON_INLINE`. That had also annoyed me a couple of times already. See e7e923f36bf20838f662f0a53c9e3203ed7f48fc. Please rebase on latest master and move your code into the `PythonCompatibility` section, after the thread state code.

**Comment by scoder** [comment]
Ah, no, the `///////// ... /////////` lines are the headers that split the section. You just cut the `FastGil` section into two. :)

I meant to move the C implementation [here](https://github.com/cython/cython/blob/e7e923f36bf20838f662f0a53c9e3203ed7f48fc/Cython/Utility/ModuleSetupCode.c#L423), without adding a new section for it. You can still keep your "TSS API" title as a normal `// ...` comment.

**Comment by sonots** [comment]
Thank you very much! It seems it is working fine now.

My remained question is: How do we define followings in pxd file?

```
#define Py_tss_NEEDS_INIT {0}
```

https://github.com/python/cpython/blob/731e18901484c75b60167a06a0ba0719a6d4827d/Include/pythread.h#L130

**Comment by sonots** [comment]
Changed to `// ...`

**Comment by scoder** [comment]
> #define Py_tss_NEEDS_INIT {0}

This is essentially an external variable of type `Py_tss_t`, i.e.

`Py_tss_t Py_tss_NEEDS_INIT`

**Comment by scoder** [comment]
Note that this new code is not currently being used anywhere. Please add a test .pyx file under `tests/run/` that exercises it. You can use a (compiled) Python function with doctests for that, which will be executed by the Python test runner. (And please ignore the comment about the `fast_gil` directive, it's not needed.)

**Comment by sonots** [comment]
Thanks! I will add tests.

**Comment by sonots** [comment]
Ah, I got following error for Py_tss_NEEDS_INIT

```
/home/sonots/.pyenv/versions/3.7-dev/include/python3.7m/pythread.h:130:31: error: expected ‘;’ before ‘}’ token
 #define Py_tss_NEEDS_INIT   {0}
                               ^
```

**Comment by scoder** [comment]
Hmm, could you try structuring it like this?
```c
#if PY_VERSION_HEX < 0x03070000
#include "pythread.h"
#ifndef Py_tss_NEEDS_INIT
#define Py_tss_NEEDS_INIT 0
typedef int Py_tss_t;
...
```

**Comment by sonots** [comment]
The error is occuring for Python 3.7-dev, so it is probably not related 🤔 ?

**Comment by scoder** [comment]
Is that the complete error output that you get or is there more? What's your C compiler?

**Comment by sonots** [comment]
The entire message is as:

```
hello.c: In function ‘PyInit_hello’:
/home/sonots/.pyenv/versions/3.7-dev/include/python3.7m/pythread.h:130:29: error: expected expression before ‘{’ token
 #define Py_tss_NEEDS_INIT   {0}
                             ^
hello.c:1181:28: note: in expansion of macro ‘Py_tss_NEEDS_INIT’
   __pyx_v_5hello_tss_key = Py_tss_NEEDS_INIT;
                            ^
/home/sonots/.pyenv/versions/3.7-dev/include/python3.7m/pythread.h:130:31: error: expected ‘;’ before ‘}’ token
 #define Py_tss_NEEDS_INIT   {0}
                               ^
hello.c:636:36: note: in definition of macro ‘__Pyx_PyBool_FromLong’
 #define __Pyx_PyBool_FromLong(b) ((b) ? __Pyx_NewRef(Py_True) : __Pyx_NewRef(Py_False))
                                    ^
hello.c:1190:38: note: in expansion of macro ‘Py_tss_NEEDS_INIT’
   __pyx_t_1 = __Pyx_PyBool_FromLong((Py_tss_NEEDS_INIT == 0)); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 15, __pyx_L1_error)
                                      ^
hello.c:1190:56: error: expected ‘)’ before ‘==’ token
   __pyx_t_1 = __Pyx_PyBool_FromLong((Py_tss_NEEDS_INIT == 0)); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 15, __pyx_L1_error)
                                                        ^
hello.c:636:36: note: in definition of macro ‘__Pyx_PyBool_FromLong’
 #define __Pyx_PyBool_FromLong(b) ((b) ? __Pyx_NewRef(Py_True) : __Pyx_NewRef(Py_False))
                                    ^
Makefile:8: recipe for target 'hello' failed
```

compiler is

```
$ gcc --version
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609
```

**Comment by scoder** [comment]
Struct initialisations are only allowed in the variable declaration. Use 
`cdef Py_tss_t tss_key = Py_tss_NEEDS_INIT`

**Comment by sonots** [comment]
Well, I am doing so ... 🤔 
https://github.com/cython/cython/pull/1939#issuecomment-337444151

**Comment by scoder** [comment]
Ah, ok. It's a global variable that Cython does not initialise statically. It should still work for a local variable. That's annoying...

Could be something to improve in CPython. *Requiring* static initialisation doesn't seem ideal.

BTW:
> The error is occurring for Python 3.7-dev, so it is probably not related
But it's still the right way to do it ;)

**Comment by sonots** [comment]
Ah, okay, so cython currently generates c codes are as follows:

```
Py_tss_t __pyx_v_tss_key;
__pyx_v_tss_key = Py_tss_NEEDS_INIT;
```

But, struct initialization must be written as:

```
Py_tss_t __pyx_v_tss_key = Py_tss_NEEDS_INIT;
```

or

```
Py_tss_t __pyx_v_tss_key;
__pyx_v_tss_key = (Py_tss_t)Py_tss_NEEDS_INIT;
```

**Comment by scoder** [comment]
If the last works, then CPython should totally use a cast in the macro.

File: Cython/Utility/ModuleSetupCode.c [hunk 1]
```diff
@@ -421,6 +421,40 @@ class __Pyx_FakeReference {
...
+#if PY_VERSION_HEX < 0x03070000 && !defined(PyThread_tss_create)
+#include "pythread.h"
+#define Py_tss_NEEDS_INIT 0
+typedef int Py_tss_t;
+static CYTHON_INLINE int PyThread_tss_create(Py_tss_t *key) {
+  *key = PyThread_create_key();
+  return 0; // PyThread_create_key reports success always
+}
+static CYTHON_INLINE Py_tss_t * PyThread_tss_alloc(void) {
+  int *key = malloc(sizeof(Py_tss_t));
```
Comment (scoder) [inline Cython/Utility/ModuleSetupCode.c]: C++ requires a cast for the `malloc()` result. I actually think that this should use `PyObject_Malloc()` and `PyObject_Free()`  instead, which is faster for small chunks of memory.

Comment (scoder) [reply Cython/Utility/ModuleSetupCode.c]: BTW, I think this should be `Py_tss_t *` and not `int *`. It's just cleaner.

**Comment by sonots** [comment]
Added tests/compile/tss.pyx, but I am not sure whether this is enough.

**Comment by scoder** [comment]
Please move the test code into a `def` function. Then you can use a doctest to run it.
Replace `mode: compile` with `mode: run` and move the file into `/tests/run/`.

**Comment by scoder** [comment]
Beautiful. Thanks!

**Comment by scoder** [comment]
Note: I promoted `Py_tss_t` to a "real" known Cython type in 8fab3aabe93c3c60b0084f30fdff6d4fd1e37b42. That avoids any static initialisation issues with `Py_tss_NEEDS_INIT` as Cython can now do that internally as part of the variable declaration.

**Comment by sonots** [comment]
Good to know, thanks!

**Pull request merged**